crates/ty_python_semantic/resources/mdtest/function/return_type.md

```diff
@@ -182,6 +182,225 @@ def f(cond: bool) -> int:
         return 2
 ```
 
+## Inferred return type
+
+### Free function
+
+If a function's return type is not annotated, it is inferred. The inferred type is the union of all
+possible return types.
+
+```py
+def f():
+    return 1
+
+reveal_type(f())  # revealed: Literal[1]
+
+def g(cond: bool):
+    if cond:
+        return 1
+    else:
+        return "a"
+
+reveal_type(g(True))  # revealed: Literal[1, "a"]
+
+# This function implicitly returns `None`.
+def h(x: int, y: str):
+    if x > 10:
+        return x
+    elif x > 5:
+        return y
+
+reveal_type(h(1, "a"))  # revealed: int | str | None
+
+def generator():
+    yield 1
+    yield 2
+    return None
+
+# TODO: Should be `Generator[Literal[1, 2], Any, None]`
+reveal_type(generator())  # revealed: None
+```
+
+The return type of a recursive function is also inferred. When the return type inference would
+diverge, it is truncated and replaced with the type `Unknown`.
+
+```py
+def fibonacci(n: int):
+    if n == 0:
+        return 0
+    elif n == 1:
+        return 1
+    else:
+        return fibonacci(n - 1) + fibonacci(n - 2)
+
+# TODO: it may be better to infer this as `int` if we can
+reveal_type(fibonacci(5))  # revealed: Literal[0, 1] | Unknown
+
+def even(n: int):
+    if n == 0:
+        return True
+    else:
+        return odd(n - 1)
+
+def odd(n: int):
+    if n == 0:
+        return False
+    else:
+        return even(n - 1)
+
+# TODO: it may be better to infer these as `bool` if we can
+reveal_type(even(1))  # revealed: bool | Unknown
```

Comment:
It's a bit unfortunate that we can't infer `bool` here, but pyright gives the same result.